### PR TITLE
Link E2E example to docs

### PR DESCRIPTION
**Pull request by avishniakov**
## Describe changes
I added sections linking E2E example into some docs sections, to help users better understand described concepts and increase adoption of ZenML using `zenml init --template`

Affected sections:
- https://app.gitbook.com/o/-MCl1Hlw9oU4xibZ-ymf/s/9oLgDi1mkUbwQiTmrEit/user-guide/advanced-guide/pipelining-features/configure-steps-pipelines#pass-any-kind-of-data-to-your-steps
- https://app.gitbook.com/o/-MCl1Hlw9oU4xibZ-ymf/s/9oLgDi1mkUbwQiTmrEit/user-guide/advanced-guide/pipelining-features/use-failure-success-hooks#defining-hooks-on-a-pipeline-level
- https://app.gitbook.com/o/-MCl1Hlw9oU4xibZ-ymf/s/9oLgDi1mkUbwQiTmrEit/user-guide/advanced-guide/pipelining-features/use-failure-success-hooks#accessing-step-information-inside-a-hook
- https://app.gitbook.com/o/-MCl1Hlw9oU4xibZ-ymf/s/9oLgDi1mkUbwQiTmrEit/user-guide/advanced-guide/pipelining-features/use-failure-success-hooks#linking-to-the-alerter-stack-component
- https://app.gitbook.com/o/-MCl1Hlw9oU4xibZ-ymf/s/9oLgDi1mkUbwQiTmrEit/user-guide/advanced-guide/pipelining-features/hyper-parameter-tuning

## Pre-requisites
Please ensure you have done the following:
- [x] I have read the **CONTRIBUTING.md** document.
- [x] If my change requires a change to docs, I have updated the documentation accordingly.
- [ ] If I have added an integration, I have updated the [integrations](https://docs.zenml.io/stacks-and-components/component-guide) table and the [corresponding website section](https://zenml.io/integrations).
- [ ] I have added tests to cover my changes.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Other (add details above)

### CODE REVIEW

File: docs/book/user-guide/advanced-guide/pipelining-features/configure-steps-pipelines.md [hunk 1]
```diff
@@ -182,6 +182,52 @@ Another way to search would be by a combination of a pipeline name where the art
 artifact = ExternalArtifact(pipeline_name="training_pipeline", artifact_name="model")
 ```
 
+<details>
+
+<summary>See it in action with the E2E example</summary>
+
+*To setup local environment used below follow recommendations from
```
Comment (bcdurak) [inline docs/book/user-guide/advanced-guide/pipelining-features/configure-steps-pipelines.md:189]:
```suggestion
*To setup the local environment used below, follow the recommendations from the
```

File: docs/book/user-guide/advanced-guide/pipelining-features/configure-steps-pipelines.md [hunk 2]
```diff
@@ -182,6 +182,52 @@ Another way to search would be by a combination of a pipeline name where the art
...
+
+<summary>See it in action with the E2E example</summary>
+
+*To setup local environment used below follow recommendations from 
+[Project templates](../../starter-guide/using-project-templates.md#advanced-guide).*
+
+In [`pipelines/batch_inference.py`](../../../../../examples/e2e/pipelines/batch_inference.py) you can find an example using `ExternalArtifact` concept to
+share Artifacts produced by training pipeline inside batch inference pipeline.
+
+On ETL stage pipeline developer can pass `sklearn.Pipeline` fitted during training for feature preprocessing and apply it to transform inference input features.
```
Comment (bcdurak) [inline docs/book/user-guide/advanced-guide/pipelining-features/configure-steps-pipelines.md:195]:
```suggestion
On the ETL stage pipeline, developers can pass a `sklearn.Pipeline` fitted during training for feature preprocessing and apply it to transform inference input features.
```

File: docs/book/user-guide/advanced-guide/pipelining-features/configure-steps-pipelines.md [hunk 3]
```diff
@@ -182,6 +182,52 @@ Another way to search would be by a combination of a pipeline name where the art
...
+With this we ensure that exactly same feature preprocessor used during training will be used during inference.
```
Comment (bcdurak) [inline docs/book/user-guide/advanced-guide/pipelining-features/configure-steps-pipelines.md:196]:
```suggestion
With this, we ensure that the exact same feature preprocessor used during training will be used during inference.
```

File: docs/book/user-guide/advanced-guide/pipelining-features/configure-steps-pipelines.md [hunk 4]
```diff
@@ -182,6 +182,52 @@ Another way to search would be by a combination of a pipeline name where the art
...
+        dataset_inf=df_inference,
+        preprocess_pipeline=ExternalArtifact(
+            pipeline_name=MetaConfig.pipeline_name_training,
+            artifact_name="preprocess_pipeline",
+        ),
+        target=target,
+    )
+```
+
+On DataQuality stage pipeline developer can pass `pd.DataFrame` used as a training dataset to be used as a reference dataset versus current inference one to apply Evidently and get Data Quality report back.
```
Comment (bcdurak) [inline docs/book/user-guide/advanced-guide/pipelining-features/configure-steps-pipelines.md:211]:
```suggestion
On the DataQuality stage pipeline, developers can pass `pd.DataFrame` used as a training dataset to be used as a reference dataset versus the current inference one to apply Evidently and get DataQuality report back.
```

File: docs/book/user-guide/advanced-guide/pipelining-features/configure-steps-pipelines.md [hunk 5]
```diff
@@ -182,6 +182,52 @@ Another way to search would be by a combination of a pipeline name where the art
...
+With this we ensure that exactly same training dataset used during training phase will be used to compare with inference dataset here.
```
Comment (bcdurak) [inline docs/book/user-guide/advanced-guide/pipelining-features/configure-steps-pipelines.md:212]:
```suggestion
With this, we ensure that the exact same training dataset used during the training phase will be used to compare with the inference dataset here.
```

File: docs/book/user-guide/advanced-guide/pipelining-features/hyper-parameter-tuning.md [hunk 6]
```diff
@@ -32,6 +32,49 @@ that would allow for a different learning rate to be used across the same
 the `select_model_step` finds which hyperparameters gave the best results or 
 performance.
 
+<details>
+
+<summary>See it in action with the E2E example</summary>
+
+*To setup local environment used below follow recommendations from
```
Comment (bcdurak) [inline docs/book/user-guide/advanced-guide/pipelining-features/hyper-parameter-tuning.md:39]:
```suggestion
*To setup the local environment used below, follow recommendations from the
```

File: docs/book/user-guide/advanced-guide/pipelining-features/hyper-parameter-tuning.md [hunk 7]
```diff
@@ -61,5 +104,24 @@ def select_model_step():
         ...
 ```
 
+<details>
+
+<summary>See it in action with the E2E example</summary>
+
+*To setup local environment used below follow recommendations from
```
Comment (bcdurak) [inline docs/book/user-guide/advanced-guide/pipelining-features/hyper-parameter-tuning.md:111]:
```suggestion
*To setup the local environment used below, follow recommendations from the
```

File: docs/book/user-guide/advanced-guide/pipelining-features/hyper-parameter-tuning.md [hunk 8]
```diff
@@ -61,5 +104,24 @@ def select_model_step():
...
 
+<details>
+
+<summary>See it in action with the E2E example</summary>
+
+*To setup local environment used below follow recommendations from 
+[Project templates](../../starter-guide/using-project-templates.md#advanced-guide).*
+
+
+In `steps/hp_tuning` folder you will find two step files, which can be used
```
Comment (bcdurak) [inline docs/book/user-guide/advanced-guide/pipelining-features/hyper-parameter-tuning.md:115]:
```suggestion
In the `steps/hp_tuning` folder, you will find two step files, which can be used 
```

File: docs/book/user-guide/advanced-guide/pipelining-features/hyper-parameter-tuning.md [hunk 9]
```diff
@@ -61,5 +104,24 @@ def select_model_step():
...
+
+*To setup local environment used below follow recommendations from 
+[Project templates](../../starter-guide/using-project-templates.md#advanced-guide).*
+
+
+In `steps/hp_tuning` folder you will find two step files, which can be used 
+as a starting point for building your own hyperparameter search tailored 
+specifically to your use case:
+- [`hp_tuning_single_search(...)`](../../../../../examples/e2e/steps/hp_tuning/hp_tuning_single_search.py) is performing a randomized 
+    search for best model hyperparameters in configured space.
```
Comment (bcdurak) [inline docs/book/user-guide/advanced-guide/pipelining-features/hyper-parameter-tuning.md:119]:
```suggestion
    search for the best model hyperparameters in a configured space.
```

File: docs/book/user-guide/advanced-guide/pipelining-features/hyper-parameter-tuning.md [hunk 10]
```diff
@@ -61,5 +104,24 @@ def select_model_step():
...
+- [`hp_tuning_select_best_model(...)`](../../../../../examples/e2e/steps/hp_tuning/hp_tuning_select_best_model.py) is searching for best hyperparameters,
+    looping other results of previous random searches to find best model according 
+    to defined metric.
```
Comment (bcdurak) [inline docs/book/user-guide/advanced-guide/pipelining-features/hyper-parameter-tuning.md:122]:
```suggestion
- [`hp_tuning_select_best_model(...)`](../../../../../examples/e2e/steps/hp_tuning/hp_tuning_select_best_model.py) is searching for the best hyperparameters,
    looping other results of previous random searches to find the best model according 
    to a defined metric.
```

File: docs/book/user-guide/advanced-guide/pipelining-features/use-failure-success-hooks.md [hunk 11]
```diff
@@ -62,6 +62,42 @@ def my_pipeline(...):
 Note, that **step-level** defined hooks take **precedence** over **pipeline-level** defined hooks.
 {% endhint %}
 
+<details>
+
+<summary>See it in action with the E2E example</summary>
+
+*To setup local environment used below follow recommendations from
```
Comment (bcdurak) [inline docs/book/user-guide/advanced-guide/pipelining-features/use-failure-success-hooks.md:69]:
```suggestion
*To setup the local environment used below, follow recommendations from the
```

File: docs/book/user-guide/advanced-guide/pipelining-features/use-failure-success-hooks.md [hunk 12]
```diff
@@ -62,6 +62,42 @@ def my_pipeline(...):
...
 {% endhint %}
 
+<details>
+
+<summary>See it in action with the E2E example</summary>
+
+*To setup local environment used below follow recommendations from 
+[Project templates](../../starter-guide/using-project-templates.md#advanced-guide).*
+
+In [`steps/alerts/notify_on.py`](../../../../../examples/e2e/steps/alerts/notify_on.py) you will find a step to notify the user
```
Comment (bcdurak) [inline docs/book/user-guide/advanced-guide/pipelining-features/use-failure-success-hooks.md:72]:
```suggestion
In [`steps/alerts/notify_on.py`](../../../../../examples/e2e/steps/alerts/notify_on.py), you will find a step to notify the user
```

File: docs/book/user-guide/advanced-guide/pipelining-features/use-failure-success-hooks.md [hunk 13]
```diff
@@ -62,6 +62,42 @@ def my_pipeline(...):
...
+about success and a function used to notify the user about step failure using the [Alerter](../../../stacks-and-components/component-guide/alerters/alerters.md)
+from the active stack.
+
+We use `@step` for success notification to only notify the user about fully successful pipeline run and not about every successful step.
+
+In [`pipelines/training.py`](../../../../../examples/e2e/pipelines/training.py) you can find the usage of a notification step and a function. We will attach
+a notify on failure function directly to the pipeline definition like this:
```
Comment (bcdurak) [inline docs/book/user-guide/advanced-guide/pipelining-features/use-failure-success-hooks.md:79]:
```suggestion
In [`pipelines/training.py`](../../../../../examples/e2e/pipelines/training.py), you can find the usage of a notification step and a function. We will attach
a `notify_on_failure` function directly to the pipeline definition like this:
```

File: docs/book/user-guide/advanced-guide/pipelining-features/use-failure-success-hooks.md [hunk 14]
```diff
@@ -62,6 +62,42 @@ def my_pipeline(...):
...
+
+```python
+@pipeline(
+    ...
+    on_failure=notify_on_failure,
+    ...
+)
+```
+
+At the very end of the training pipeline we will execute the notify on success step, but only after all other steps finished - we control it with `after` statement as follows:
```
Comment (bcdurak) [inline docs/book/user-guide/advanced-guide/pipelining-features/use-failure-success-hooks.md:89]:
```suggestion
At the very end of the training pipeline, we will execute the `notify_on_success` step, but only after all other steps have finished - we control it with `after` statement as follows:
```

File: docs/book/user-guide/advanced-guide/pipelining-features/use-failure-success-hooks.md [hunk 15]
```diff
@@ -84,6 +120,50 @@ def my_step(some_parameter: int = 1)
     raise ValueError("My exception")
 ```
 
+<details>
+
+<summary>See it in action with the E2E example</summary>
+
+*To setup local environment used below follow recommendations from
```
Comment (bcdurak) [inline docs/book/user-guide/advanced-guide/pipelining-features/use-failure-success-hooks.md:127]:
```suggestion
*To setup the local environment used below follow recommendations from the
```

File: docs/book/user-guide/advanced-guide/pipelining-features/use-failure-success-hooks.md [hunk 16]
```diff
@@ -84,6 +120,50 @@ def my_step(some_parameter: int = 1)
...
 ```
 
+<details>
+
+<summary>See it in action with the E2E example</summary>
+
+*To setup local environment used below follow recommendations from 
+[Project templates](../../starter-guide/using-project-templates.md#advanced-guide).*
+
+In [`steps/alerts/notify_on.py`](../../../../../examples/e2e/steps/alerts/notify_on.py) you will find a step to notify the user
```
Comment (bcdurak) [inline docs/book/user-guide/advanced-guide/pipelining-features/use-failure-success-hooks.md:130]:
```suggestion
In [`steps/alerts/notify_on.py`](../../../../../examples/e2e/steps/alerts/notify_on.py), you will find a step to notify the user
```

File: docs/book/user-guide/advanced-guide/pipelining-features/use-failure-success-hooks.md [hunk 17]
```diff
@@ -62,6 +62,42 @@ def my_pipeline(...):
...
+about success and a function used to notify the user about step failure using the [Alerter](../../../stacks-and-components/component-guide/alerters/alerters.md)
+from the active stack.
+
+We use `@step` for success notification to only notify the user about fully successful pipeline run and not about every successful step.
```
Comment (bcdurak) [inline docs/book/user-guide/advanced-guide/pipelining-features/use-failure-success-hooks.md:76]:
```suggestion
We use `@step` for success notification to only notify the user about a fully successful pipeline run and not about every successful step.
```

File: docs/book/user-guide/advanced-guide/pipelining-features/use-failure-success-hooks.md [hunk 18]
```diff
@@ -84,6 +120,50 @@ def my_step(some_parameter: int = 1)
...
```
Comment (bcdurak) [inline docs/book/user-guide/advanced-guide/pipelining-features/use-failure-success-hooks.md:134]:
```suggestion
We use `@step` for success notification to only notify the user about a fully successful pipeline run and not about every successful step.
```

File: docs/book/user-guide/advanced-guide/pipelining-features/use-failure-success-hooks.md [hunk 19]
```diff
@@ -84,6 +120,50 @@ def my_step(some_parameter: int = 1)
...
+
+Inside the helper function `build_message()` you will find an example how developers can work with [StepContext](fetch-metadata-within-steps.md) to form a proper notification:
```
Comment (bcdurak) [inline docs/book/user-guide/advanced-guide/pipelining-features/use-failure-success-hooks.md:136]:
```suggestion
Inside the helper function `build_message()`, you will find an example on how developers can work with [StepContext](fetch-metadata-within-steps.md) to form a proper notification:
```

File: docs/book/user-guide/advanced-guide/pipelining-features/use-failure-success-hooks.md [hunk 20]
```diff
@@ -108,6 +188,54 @@ def my_step(...):
     ...
 ```
 
+<details>
+
+<summary>See it in action with the E2E example</summary>
+
+*To setup local environment used below follow recommendations from
```
Comment (bcdurak) [inline docs/book/user-guide/advanced-guide/pipelining-features/use-failure-success-hooks.md:195]:
```suggestion
*To setup the local environment used below follow recommendations from the
```

File: docs/book/user-guide/advanced-guide/pipelining-features/use-failure-success-hooks.md [hunk 21]
```diff
@@ -108,6 +188,54 @@ def my_step(...):
...
 ```
 
+<details>
+
+<summary>See it in action with the E2E example</summary>
+
+*To setup local environment used below follow recommendations from 
+[Project templates](../../starter-guide/using-project-templates.md#advanced-guide).*
+
+In [`steps/alerts/notify_on.py`](../../../../../examples/e2e/steps/alerts/notify_on.py) you will find a step to notify the user
```
Comment (bcdurak) [inline docs/book/user-guide/advanced-guide/pipelining-features/use-failure-success-hooks.md:198]:
```suggestion
In [`steps/alerts/notify_on.py`](../../../../../examples/e2e/steps/alerts/notify_on.py), you will find a step to notify the user
```

File: docs/book/user-guide/advanced-guide/pipelining-features/use-failure-success-hooks.md [hunk 22]
```diff
@@ -108,6 +188,54 @@ def my_step(...):
...
+about success and a function used to notify the user about step failure using the [Alerter](../../../stacks-and-components/component-guide/alerters/alerters.md)
+from the active stack.
+
+We use `@step` for success notification to only notify the user about fully successful pipeline run and not about every successful step.
```
Comment (bcdurak) [inline docs/book/user-guide/advanced-guide/pipelining-features/use-failure-success-hooks.md:202]:
```suggestion
We use `@step` for success notification to only notify the user about a fully successful pipeline run and not about every successful step.
```

File: docs/book/user-guide/advanced-guide/pipelining-features/use-failure-success-hooks.md [hunk 23]
```diff
@@ -108,6 +188,54 @@ def my_step(...):
...
+
+Inside code file you can find how developers can work with Al component to send notification messages across configured channels:
```
Comment (bcdurak) [inline docs/book/user-guide/advanced-guide/pipelining-features/use-failure-success-hooks.md:204]:
```suggestion
Inside this code file, you can find how developers can work with Al component to send notification messages across configured channels:
```

File: docs/book/user-guide/starter-guide/follow-best-practices.md [hunk 24]
```diff
@@ -34,6 +34,8 @@ Until now, you probably have kept all your code in one single file. In productio
 └── run.py
 ```
 
+Check out how to initialize your project from template following best practices in [Project templates](./using-project-templates.md#generating-project-from-a-project-template) section.
```
Comment (bcdurak) [inline docs/book/user-guide/starter-guide/follow-best-practices.md:37]:
```suggestion
Check out how to initialize your project from a template following best practices in the [Project templates](./using-project-templates.md#generating-project-from-a-project-template) section.
```

File: docs/book/user-guide/starter-guide/using-project-templates.md [hunk 25]
```diff
@@ -36,4 +36,32 @@ zenml init --template <short_name_of_template> --template-with-defaults
...
+cd starter
+zenml init --template starter --template-with-defaults
+``` -->
+
+### Advanced Guide
+
+Our Advanced Guide documentation is built around the `E2E Batch` project template codes. 
+Mostly examples will come from it, so we highly recommend you to install `e2e_batch` template
+with `--template-with-defaults` flag before diving deeper into this documentation section, 
+so you can follow this guide along using your own local environment.
```
Comment (bcdurak) [inline docs/book/user-guide/starter-guide/using-project-templates.md:59]:
```suggestion
Our Advanced Guide documentation is built around the `E2E Batch` project template codes. 
Most examples will be based on it, so we highly recommend you to install the `e2e_batch` template
with `--template-with-defaults` flag before diving deeper into this documentation section, 
so you can follow this guide along using your own local environment.
```

File: docs/book/user-guide/starter-guide/using-project-templates.md [hunk 26]
```diff
@@ -36,4 +36,32 @@ zenml init --template <short_name_of_template> --template-with-defaults
 # example: zenml init --template e2e_batch --template-with-defaults
 ```
 
+## Project templates used in our examples
+
+<!-- ### Starter Guide
+
+Our Starter Guide documentation is built around the `Starter` project template codes. 
+Mostly examples will come from it, so we highly recommend you to install `starter` template
```
Comment (bcdurak) [inline docs/book/user-guide/starter-guide/using-project-templates.md:44]:
```suggestion
Most examples will be based on it, so we highly recommend you to install the `starter` template
```

File: docs/book/user-guide/advanced-guide/pipelining-features/configure-steps-pipelines.md [hunk 27]
```diff
@@ -182,6 +182,52 @@ Another way to search would be by a combination of a pipeline name where the art
...
 
+<details>
+
+<summary>See it in action with the E2E example</summary>
+
+*To setup local environment used below follow recommendations from 
+[Project templates](../../starter-guide/using-project-templates.md#advanced-guide).*
+
+In [`pipelines/batch_inference.py`](../../../../../examples/e2e/pipelines/batch_inference.py) you can find an example using `ExternalArtifact` concept to
+share Artifacts produced by training pipeline inside batch inference pipeline.
```
Comment (bcdurak) [inline docs/book/user-guide/advanced-guide/pipelining-features/configure-steps-pipelines.md:193]:
```suggestion
In [`pipelines/batch_inference.py`](../../../../../examples/e2e/pipelines/batch_inference.py), you can find an example using the `ExternalArtifact` concept to
share Artifacts produced by a training pipeline inside a batch inference pipeline.
```

File: docs/book/user-guide/advanced-guide/pipelining-features/hyper-parameter-tuning.md [hunk 28]
```diff
@@ -32,6 +32,49 @@ that would allow for a different learning rate to be used across the same
...
+
+<summary>See it in action with the E2E example</summary>
+
+*To setup local environment used below follow recommendations from 
+[Project templates](../../starter-guide/using-project-templates.md#advanced-guide).*
+
+In [`pipelines/training.py`](../../../../../examples/e2e/pipelines/training.py) you will find a training pipeline with 
+`Hyperparameter tuning stage` section. It contains a `for` loop over configured 
+model search spaces to run `hp_tuning_single_search` on followed by 
+`hp_tuning_select_best_model` executed after all search steps are completed. As
```
Comment (bcdurak) [inline docs/book/user-guide/advanced-guide/pipelining-features/hyper-parameter-tuning.md:45]:
```suggestion
In [`pipelines/training.py`](../../../../../examples/e2e/pipelines/training.py), you will find a training pipeline with a
`Hyperparameter tuning stage` section. It contains a `for` loop that runs the `hp_tuning_single_search` over the configured model search spaces, followed by the
`hp_tuning_select_best_model` being executed after all search steps are completed. As 
```

**Review by bcdurak — CHANGES_REQUESTED** [review]
I found a few typos and tried to leave suggestions as much as I could. Hope it helps :)

File: docs/book/user-guide/advanced-guide/pipelining-features/configure-steps-pipelines.md (hunk 1 above)
Comment (avishniakov) [reply docs/book/user-guide/advanced-guide/pipelining-features/configure-steps-pipelines.md:189]: Will update via commit in all places.

**Comment by avishniakov** [comment]
> I found a few typos and tried to leave suggestions as much as I could. Hope it helps :)

Oh man, that's huge - thanks! I tried to clean it up with tools before committing, but seems I need to build up my practices with docs 🙂

**Review by bcdurak — APPROVED** [review]

**Pull request merged**